aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,148 @@
+module aptos_framework::nonce_validation {
+    use aptos_std::table::{Self, Table};
+    use aptos_std::timestamp;
+    use aptos_std::big_ordered_map::{Self, BigOrderedMap};
+    use aptos_std::aptos_hash::sip_hash_from_value;
+    use aptos_std::option;
+    use aptos_std::vector;
+    friend aptos_framework::genesis;
+    friend aptos_framework::transaction_validation;
+
+    const NUM_BUCKETS: u64 = 50000;
+    const MAP_SWITCH_TIME: u64 = 75;
+    const MAX_EXPIRATION_TIME: u64 = 60;
+
+    struct Bucket has store {
+        // Maps NonceKey = (address, nonce) to txn expiration time.
+        // Using multiple maps to facilitate easier garbage collection.
+        nonces: vector<BigOrderedMap<NonceKey, u64>>,
+        // last_stored_times[i] stores the timestamp at which a nonce entry 
+        // is last added to the nonces[i] map.
+        last_stored_times: vector<u64>,
+    }
+
+    struct NonceKey has copy, drop, store {
+        sender_address: address,
+        nonce: u64,
+    }
+
+    struct NonceHistory has key {
+        // Key = sip_hash(NonceKey) % NUM_BUCKETS
+        // Value = Bucket
+        nonce_table: Table<u64, Bucket>,
+        // Used to facilitate prefill the nonce_table with empty buckets
+        // one by one using `add_nonce_bucket` method.
+        // This is the next_key to prefill with an empty bucket
+        next_key: u64,
+    }
+
+    public(friend) fun initialize(aptos_framework: &signer) {
+        initialize_nonce_table(aptos_framework);
+    }
+
+    public entry fun initialize_nonce_table(aptos_framework: &signer) {
+        if (!exists<NonceHistory>(@aptos_framework)) {
+            let table = table::new();
+            let nonce_history = NonceHistory {
+                nonce_table: table,
+                next_key: 0,
+            };
+            move_to<NonceHistory>(aptos_framework, nonce_history);
+        };
+    }
+
+    // This method is used to prefill the nonce_table with empty buckets one by one.
+    public entry fun add_nonce_bucket(aptos_framework: &signer) acquires NonceHistory {
+        if (exists<NonceHistory>(@aptos_framework)) {
+            let current_time = timestamp::now_seconds();
+            let nonce_history = borrow_global_mut<NonceHistory>(@aptos_framework);
+            if (!table::contains(&nonce_history.nonce_table, nonce_history.next_key)) {
+                let nonces = vector::empty();
+                let last_stored_times = vector::empty();
+                vector::push_back(&mut nonces, big_ordered_map::new());
+                vector::push_back(&mut nonces, big_ordered_map::new());
+                vector::push_back(&mut last_stored_times, 0);
+                vector::push_back(&mut last_stored_times, 0);
```

Comment:
could you refactor all this with move 2?

Reply:
Done.